File: 31/python/user_input.py
#!/usr/bin/env python
# coding: -utf8

import sys


# readUserInput プロンプトを出してユーザー入力を取得する.
# EOFが入力された場合はプログラム終了する
def read_user_input(prompter):
    try:
        return input(prompter)
    except EOFError:
        sys.exit(1)


# readUserInputWithValidator プロンプトを出してユーザー入力を取得して,
# 希望の型に変換する.希望に沿わない入力だった場合は再度入力をうながす.
def read_user_input_with_validator(prompter, validator, invalidMsg):

    while True:
        user_input = read_user_input(prompter)
        try:
            return validator(user_input)
        except ValueError:
            print(invalidMsg)
            continue

File: 31/go/user_input.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var scanner *bufio.Scanner

func init() {
	scanner = bufio.NewScanner(os.Stdin)
}

func input(msg string) (string, error) {
	_, err := fmt.Printf(msg)
	if err != nil {
		return "", err
	}
	if !scanner.Scan() {
		if scanner.Err() == nil {
			// EOF
			return "", fmt.Errorf("interrupted by EOF")
		}
		// ERROR
		return "", scanner.Err()
	}
	return scanner.Text(), nil
}

func readUserInput(prompter, invalidMsg string) (int, error) {

retry:
	userInput, err := input(prompter)
	if err != nil {
		// EOF or ERROR
		return 0, err
	}
	intValue, err := strconv.Atoi(userInput)
	if err != nil {
		fmt.Println(invalidMsg)
		goto retry
	}
	return intValue, nil
}
